In [1]:
import pandas as pd

In [2]:
orders = pd.read_csv('data/olist_orders_dataset.csv')
customers = pd.read_csv('data/olist_customers_dataset.csv')
order_items = pd.read_csv('data/olist_order_items_dataset.csv')
products = pd.read_csv('data/olist_products_dataset.csv')

In [3]:
tables = {
    'orders': orders,
    'customers': customers,
    'order_items': order_items,
    'products': products
}

In [4]:
for table_name, df in tables.items():
    print(table_name, df.shape)

orders (99441, 8)
customers (99441, 5)
order_items (112650, 7)
products (32951, 9)


In [5]:
for table_name, df in tables.items():
    print(f'\n===== {table_name.upper()} =====')
    
    print('\nSHAPE:')
    print(df.shape)
    
    print('\nHEAD:')
    display(df.head())
    
    print('\nDTYPES:')
    display(df.dtypes)
    
    print('\nINFO:')
    df.info()
    
    print('\nMISSING VALUES:')
    display(df.isna().sum())
    
    print('\nFULL DUPLICATES:')
    print(df.duplicated().sum())


===== ORDERS =====

SHAPE:
(99441, 8)

HEAD:


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00



DTYPES:


order_id                         str
customer_id                      str
order_status                     str
order_purchase_timestamp         str
order_approved_at                str
order_delivered_carrier_date     str
order_delivered_customer_date    str
order_estimated_delivery_date    str
dtype: object


INFO:
<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype
---  ------                         --------------  -----
 0   order_id                       99441 non-null  str  
 1   customer_id                    99441 non-null  str  
 2   order_status                   99441 non-null  str  
 3   order_purchase_timestamp       99441 non-null  str  
 4   order_approved_at              99281 non-null  str  
 5   order_delivered_carrier_date   97658 non-null  str  
 6   order_delivered_customer_date  96476 non-null  str  
 7   order_estimated_delivery_date  99441 non-null  str  
dtypes: str(8)
memory usage: 6.1 MB

MISSING VALUES:


order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64


FULL DUPLICATES:
0

===== CUSTOMERS =====

SHAPE:
(99441, 5)

HEAD:


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP



DTYPES:


customer_id                   str
customer_unique_id            str
customer_zip_code_prefix    int64
customer_city                 str
customer_state                str
dtype: object


INFO:
<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   customer_id               99441 non-null  str  
 1   customer_unique_id        99441 non-null  str  
 2   customer_zip_code_prefix  99441 non-null  int64
 3   customer_city             99441 non-null  str  
 4   customer_state            99441 non-null  str  
dtypes: int64(1), str(4)
memory usage: 3.8 MB

MISSING VALUES:


customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64


FULL DUPLICATES:
0

===== ORDER_ITEMS =====

SHAPE:
(112650, 7)

HEAD:


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14



DTYPES:


order_id                   str
order_item_id            int64
product_id                 str
seller_id                  str
shipping_limit_date        str
price                  float64
freight_value          float64
dtype: object


INFO:
<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  str    
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  str    
 3   seller_id            112650 non-null  str    
 4   shipping_limit_date  112650 non-null  str    
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), str(4)
memory usage: 6.0 MB

MISSING VALUES:


order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64


FULL DUPLICATES:
0

===== PRODUCTS =====

SHAPE:
(32951, 9)

HEAD:


,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0



DTYPES:


product_id                        str
product_category_name             str
product_name_lenght           float64
product_description_lenght    float64
product_photos_qty            float64
product_weight_g              float64
product_length_cm             float64
product_height_cm             float64
product_width_cm              float64
dtype: object


INFO:
<class 'pandas.DataFrame'>
RangeIndex: 32951 entries, 0 to 32950
Data columns (total 9 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   product_id                  32951 non-null  str    
 1   product_category_name       32341 non-null  str    
 2   product_name_lenght         32341 non-null  float64
 3   product_description_lenght  32341 non-null  float64
 4   product_photos_qty          32341 non-null  float64
 5   product_weight_g            32949 non-null  float64
 6   product_length_cm           32949 non-null  float64
 7   product_height_cm           32949 non-null  float64
 8   product_width_cm            32949 non-null  float64
dtypes: float64(7), str(2)
memory usage: 2.3 MB

MISSING VALUES:


product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64


FULL DUPLICATES:
0


In [6]:
orders_clean = orders.copy()
customers_clean = customers.copy()
order_items_clean = order_items.copy()
products_clean = products.copy()

In [7]:
delivered_orders = orders_clean.loc[ orders_clean['order_status'] == 'delivered' ]

In [8]:
delivered_orders_selected = orders_clean.loc[ orders_clean['order_status'] == 'delivered',
    ['order_id',
    'customer_id',
    'order_status',
    'order_purchase_timestamp',
    'order_delivered_customer_date']
 ]

In [9]:
problem_orders = orders_clean.loc[ (orders_clean['order_status'] == 'canceled') | ( orders_clean['order_status'] == 'unavailable')]

In [10]:
problem_orders['order_status'].value_counts()

order_status
canceled       625
unavailable    609
Name: count, dtype: int64

In [11]:
orders_clean.iloc[:10, :4]

,order_id,customer_id,order_status,order_purchase_timestamp
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39
5,a4591c265e18cb1dcee52889e2d8acc3,503740e9ca751ccdda7ba28e9ab8f608,delivered,2017-07-09 21:57:05
6,136cce7faa42fdb2cefd53fdc79a6098,ed0271e0b7da060a393796590e7b737a,invoiced,2017-04-11 12:22:08
7,6514b8ad8028c9f2cc2374ded245783f,9bdf08b4b3b52b5526ff42d37d47f222,delivered,2017-05-16 13:10:30
8,76c6e866289321a7c93b82b54852dc33,f54a9f0e6b351c431402b8461ea51999,delivered,2017-01-23 18:29:09
9,e69bfb5eb88e0ed6a785585b27e16dbf,31ad1d1b63eb9962463f764d4e6e0c9d,delivered,2017-07-29 11:55:02


In [12]:
orders_clean.iloc[(-5):]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
99436,9c5dedf39a927c1b2549525ed64a053c,39bd1228ee8140590ac3aca26f2dfe00,delivered,2017-03-09 09:54:05,2017-03-09 09:54:05,2017-03-10 11:18:03,2017-03-17 15:08:01,2017-03-28 00:00:00
99437,63943bddc261676b46f01ca7ac2f7bd8,1fca14ff2861355f6e5f14306ff977a7,delivered,2018-02-06 12:58:58,2018-02-06 13:10:37,2018-02-07 23:22:42,2018-02-28 17:37:56,2018-03-02 00:00:00
99438,83c1379a015df1e13d02aae0204711ab,1aa71eb042121263aafbe80c1b562c9c,delivered,2017-08-27 14:46:43,2017-08-27 15:04:16,2017-08-28 20:52:26,2017-09-21 11:24:17,2017-09-27 00:00:00
99439,11c177c8e97725db2631073c19f07b62,b331b74b18dc79bcdf6532d51e1637c1,delivered,2018-01-08 21:28:27,2018-01-08 21:36:21,2018-01-12 15:35:03,2018-01-25 23:32:54,2018-02-15 00:00:00
99440,66dea50a8b16d9b4dee7af250b4be1a5,edb027a75a1449115f6b43211ae02a24,delivered,2018-03-08 20:57:30,2018-03-09 11:20:28,2018-03-09 22:11:59,2018-03-16 13:08:30,2018-04-03 00:00:00


In [13]:
orders_clean.iloc[[0, 10, 20], [0, 2, 4]]

,order_id,order_status,order_approved_at
0,e481f51cbdc54678b7cc49136f2d6af7,delivered,2017-10-02 11:07:15
10,e6ce16cb79ec1d90b1da9085a6118aeb,delivered,2017-05-16 19:50:18
20,203096f03d82e0dffbc41ebc2e2bcfb7,delivered,2017-09-19 04:04:09


high_price = 500
high_freight = 100

Создай четыре выборки:

Из orders_clean — заказы со статусом delivered или shipped, используя |.
Из orders_clean — все недоставленные заказы, используя ~ и .isin().
Из order_items_clean — позиции с price > high_price.
Из order_items_clean — позиции одновременно с price > high_price и freight_value > high_freight, используя &.

Все четыре выражения напиши в одной ячейке и сохрани результаты в отдельные переменные. Это последнее задание этапа 1.

In [14]:
mask = (orders_clean['order_status'] == 'delivered') | (orders_clean['order_status'] == 'shipped')
result = orders_clean.loc[mask]
result

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00
...,...,...,...,...,...,...,...,...
99436,9c5dedf39a927c1b2549525ed64a053c,39bd1228ee8140590ac3aca26f2dfe00,delivered,2017-03-09 09:54:05,2017-03-09 09:54:05,2017-03-10 11:18:03,2017-03-17 15:08:01,2017-03-28 00:00:00
99437,63943bddc261676b46f01ca7ac2f7bd8,1fca14ff2861355f6e5f14306ff977a7,delivered,2018-02-06 12:58:58,2018-02-06 13:10:37,2018-02-07 23:22:42,2018-02-28 17:37:56,2018-03-02 00:00:00
99438,83c1379a015df1e13d02aae0204711ab,1aa71eb042121263aafbe80c1b562c9c,delivered,2017-08-27 14:46:43,2017-08-27 15:04:16,2017-08-28 20:52:26,2017-09-21 11:24:17,2017-09-27 00:00:00
99439,11c177c8e97725db2631073c19f07b62,b331b74b18dc79bcdf6532d51e1637c1,delivered,2018-01-08 21:28:27,2018-01-08 21:36:21,2018-01-12 15:35:03,2018-01-25 23:32:54,2018-02-15 00:00:00


In [15]:
mask = ~orders_clean['order_status'].isin(['delivered'])
result = orders_clean[mask]
result

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
6,136cce7faa42fdb2cefd53fdc79a6098,ed0271e0b7da060a393796590e7b737a,invoiced,2017-04-11 12:22:08,2017-04-13 13:25:17,NaN,NaN,2017-05-09 00:00:00
44,ee64d42b8cf066f35eac1cf57de1aa85,caded193e8e47b8362864762a83db3c5,shipped,2018-06-04 16:44:48,2018-06-05 04:31:18,2018-06-05 14:32:00,NaN,2018-06-28 00:00:00
103,0760a852e4e9d89eb77bf631eaaf1c84,d2a79636084590b7465af8ab374a8cf5,invoiced,2018-08-03 17:44:42,2018-08-07 06:15:14,NaN,NaN,2018-08-21 00:00:00
128,15bed8e2fec7fdbadb186b57c46c92f2,f3f0e613e0bdb9c7cee75504f0f90679,processing,2017-09-03 14:22:03,2017-09-03 14:30:09,NaN,NaN,2017-10-03 00:00:00
154,6942b8da583c2f9957e990d028607019,52006a9383bf149a4fb24226b173106f,shipped,2018-01-10 11:33:07,2018-01-11 02:32:30,2018-01-11 19:39:23,NaN,2018-02-07 00:00:00
...,...,...,...,...,...,...,...,...
99283,3a3cddda5a7c27851bd96c3313412840,0b0d6095c5555fe083844281f6b093bb,canceled,2018-08-31 16:13:44,NaN,NaN,NaN,2018-10-01 00:00:00
99313,e9e64a17afa9653aacf2616d94c005b8,b4cd0522e632e481f8eaf766a2646e86,processing,2018-01-05 23:07:24,2018-01-09 07:18:05,NaN,NaN,2018-02-06 00:00:00
99347,a89abace0dcc01eeb267a9660b5ac126,2f0524a7b1b3845a1a57fcf3910c4333,canceled,2018-09-06 18:45:47,NaN,NaN,NaN,2018-09-27 00:00:00
99348,a69ba794cc7deb415c3e15a0a3877e69,726f0894b5becdf952ea537d5266e543,unavailable,2017-08-23 16:28:04,2017-08-28 15:44:47,NaN,NaN,2017-09-15 00:00:00


In [16]:
high_price = 500
mask = order_items_clean['price'] > high_price
result = order_items_clean.loc[mask]
result

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
7,000576fe39319847cbb9d288c5617fa6,1,557d850972a7d6f792fd18ae1400d9b6,5996cddab893a4652a15592fb58ab8db,2018-07-10 12:30:45,810.0,70.75
16,0009c9a17f916a706d71784483a5d643,1,3f27ac8e699df3d300ec4a5d8c5cf0b2,fcb5ace8bcc92f75707dc0f01a27d269,2018-05-02 09:31:53,639.0,11.34
39,0017afd5076e074a48f1f1a4c7bac9c5,1,fe59a1e006df3ac42bf0ceb876d70969,25c5c91f63607446a97b143d2d535d31,2017-04-24 22:25:19,809.1,44.29
158,006127b8b9a1681a982313ed7129c3c0,1,f819f0c84a64f02d3a5606ca95edd272,7d13fca15225358621be4086e1eb0964,2018-03-15 02:30:23,540.0,16.22
265,00ab9fa91aafa3d881164b0da48999aa,1,224037b21bfa8621e39c265513ee6594,ceaec5548eefc6e23e6607c5435102e7,2018-06-04 09:50:04,579.0,65.34
...,...,...,...,...,...,...,...
112449,ff85c3e4329457e83cb474799a257ccc,1,144266e1585320d057797a369596bb8b,850f4f8af5ea87287ac68de36e29107f,2018-01-24 16:57:45,899.0,17.50
112553,ffc2638415f3ce34e88641eef792c1fc,1,bc2d6310e5cbd60facb2f9c8d9471597,dc317f341ab0e22f39acbd9dbf9b4a1f,2017-06-12 13:41:58,629.0,67.08
112557,ffc49b9c9d666410e9ecdb0a3efadb9c,1,f96ff9b997681965689f78bef092fdd9,2a5b78b41cd05baeac8df54c6606b92c,2018-06-05 19:56:23,765.0,27.06
112571,ffd5e80bfdaf69504efd7562a3d522dd,1,bd67e67dcfcc53e712a29c215ec26de5,3d39fbf22a665714baf62fae36029e23,2017-08-28 09:46:08,584.9,16.65


In [17]:
high_price = 500
high_freight = 100
mask = (order_items_clean['price'] > high_price) &  (order_items_clean['freight_value'] > high_freight)
result = order_items_clean.loc[mask]
result

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
865,02014f2495eef0e869616829d481d743,1,43cc8e4d981bc04b9d78b12e8a908d41,6061155addc1e54b4cfb51c1c2a32ad8,2018-08-24 11:05:25,1240.00,102.63
1084,028558cddca332b40b4ac4cb77407edf,1,a233df9a388d27dbdfd31731d4236db0,784ba75dd9d20200c4caed3d7a77141a,2017-07-04 10:35:09,2649.99,134.17
1122,029c545413b10a6bd7540401f6eef0a3,1,8d4e92265a16e69a1e1d76e67e46d72f,54219883e72aad869adfb2a54b7bfa0f,2018-07-17 09:30:15,1350.00,294.76
1751,03ffce741877a000ca0fb3520673d4c4,1,660422061e06da17ca6101e9d6b7aae8,76d5af76d0271110f9af36c92573f765,2016-10-10 16:02:37,649.00,169.12
2018,04911e590c1c5c9b8e750d3ac1344d54,1,ab495f166205a883ffe5ab0b5b55f867,01cf7e3d21494c41fb86034f2e714fa1,2017-08-08 15:15:16,629.90,117.35
...,...,...,...,...,...,...,...
111245,fcbb6af360b31b05460c2c8e524588c0,1,97c948ebc8c04b26b7bbb095d4228f2a,17f51e7198701186712e53a39c564617,2017-12-18 14:57:29,1050.00,144.38
111595,fd7a23284efc47d1c49b2c1cc6ed4cb4,1,6a459bd9608a1b384d6cb4c79aaa83e2,712e6ed8aa4aa1fa65dab41fed5737e4,2018-03-09 10:10:56,669.00,118.53
111842,fe12e676fd493885d417a34bc3917411,1,1d16c0509e7011656e03c6a582ff6667,17f51e7198701186712e53a39c564617,2017-06-06 18:35:15,760.00,312.41
111957,fe579bd413f722e51bc98182441e135c,1,4aa52adb5489aa151e17023d4e153297,f45122a9ab94eb4f3f8953578bc0c560,2017-05-17 17:10:13,1878.33,102.38


In [18]:
date_columns = [
    'order_purchase_timestamp',
    'order_approved_at',
    'order_delivered_carrier_date',
    'order_delivered_customer_date',
    'order_estimated_delivery_date'
]

for column in date_columns:
    orders_clean[column] = pd.to_datetime(orders_clean[column],
    errors = 'coerce')

In [19]:
orders_clean[date_columns].dtypes

order_purchase_timestamp         datetime64[us]
order_approved_at                datetime64[us]
order_delivered_carrier_date     datetime64[us]
order_delivered_customer_date    datetime64[us]
order_estimated_delivery_date    datetime64[us]
dtype: object

In [20]:
print(orders_clean[date_columns].isna().sum())

order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64


In [21]:
orders_clean[date_columns].agg(['min', 'max'])

,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
min,2016-09-04 21:15:19,2016-09-15 12:16:38,2016-10-08 10:34:01,2016-10-11 13:46:32,2016-09-30
max,2018-10-17 17:30:18,2018-09-03 17:40:06,2018-09-11 19:48:28,2018-10-17 13:22:46,2018-11-12


In [22]:
for column in date_columns:
    print(f'\n===== {column} =====')
    missing_by_status = orders_clean.loc[orders_clean[column].isna(), 'order_status'].value_counts()
    print(missing_by_status)


===== order_purchase_timestamp =====
Series([], Name: count, dtype: int64)

===== order_approved_at =====
order_status
canceled     141
delivered     14
created        5
Name: count, dtype: int64

===== order_delivered_carrier_date =====
order_status
unavailable    609
canceled       550
invoiced       314
processing     301
created          5
approved         2
delivered        2
Name: count, dtype: int64

===== order_delivered_customer_date =====
order_status
shipped        1107
canceled        619
unavailable     609
invoiced        314
processing      301
delivered         8
created           5
approved          2
Name: count, dtype: int64

===== order_estimated_delivery_date =====
Series([], Name: count, dtype: int64)


In [23]:
date_columns  = [
    'order_purchase_timestamp',
    'order_approved_at',
    'order_delivered_carrier_date',
    'order_delivered_customer_date',
    'order_estimated_delivery_date'
]

for column in date_columns:
    orders_clean[column] = pd.to_datetime(orders_clean[column],
    errors = 'coerce')

In [24]:
order_items_clean[
    ['price', 'freight_value']
].dtypes

price            float64
freight_value    float64
dtype: object

In [25]:
products_clean[
    [
        'product_weight_g',
        'product_length_cm',
        'product_height_cm',
        'product_width_cm',
        'product_photos_qty'
    ]
].dtypes

product_weight_g      float64
product_length_cm     float64
product_height_cm     float64
product_width_cm      float64
product_photos_qty    float64
dtype: object

In [26]:
(order_items_clean[['price', 'freight_value']] < 0).sum()

price            0
freight_value    0
dtype: int64

In [27]:
(products_clean[[
        'product_weight_g',
        'product_length_cm',
        'product_height_cm',
        'product_width_cm',
        'product_photos_qty'
    ]] < 0).sum()

product_weight_g      0
product_length_cm     0
product_height_cm     0
product_width_cm      0
product_photos_qty    0
dtype: int64

In [28]:
(order_items_clean[['price', 'freight_value']] == 0).sum()

price              0
freight_value    383
dtype: int64

In [29]:
(products_clean[[
        'product_weight_g',
        'product_length_cm',
        'product_height_cm',
        'product_width_cm',
        'product_photos_qty'
    ]] == 0).sum()

product_weight_g      4
product_length_cm     0
product_height_cm     0
product_width_cm      0
product_photos_qty    0
dtype: int64

Условие: через .loc выбери из products_clean все строки, где:

product_weight_g == 0

In [30]:
zero_weight_products = products_clean.loc[products_clean['product_weight_g'] == 0]
zero_weight_products


,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
9769,81781c0fed9fe1ad6e8c81fca1e1cb08,cama_mesa_banho,51.0,529.0,1.0,0.0,30.0,25.0,30.0
13683,8038040ee2a71048d4bdbbdc985b69ab,cama_mesa_banho,48.0,528.0,1.0,0.0,30.0,25.0,30.0
14997,36ba42dd187055e1fbe943b2d11430ca,cama_mesa_banho,53.0,528.0,1.0,0.0,30.0,25.0,30.0
32079,e673e90efa65a5409ff4196c038bb5af,cama_mesa_banho,53.0,528.0,1.0,0.0,30.0,25.0,30.0


In [31]:
order_items_clean[
    ['price', 'freight_value']
].isna().sum()

price            0
freight_value    0
dtype: int64

In [32]:
products_clean[[
        'product_weight_g',
        'product_length_cm',
        'product_height_cm',
        'product_width_cm',
        'product_photos_qty'
    ]].isna().sum()

product_weight_g        2
product_length_cm       2
product_height_cm       2
product_width_cm        2
product_photos_qty    610
dtype: int64

In [33]:
order_items_clean[
    ['price', 'freight_value']
].describe()

,price,freight_value
count,112650.000000,112650.000000
mean,120.653739,19.990320
std,183.633928,15.806405
min,0.850000,0.000000
25%,39.900000,13.080000
50%,74.990000,16.260000
75%,134.900000,21.150000
max,6735.000000,409.680000


In [34]:
products_clean[[
        'product_weight_g',
        'product_length_cm',
        'product_height_cm',
        'product_width_cm',
        'product_photos_qty'
    ]].describe()

,product_weight_g,product_length_cm,product_height_cm,product_width_cm,product_photos_qty
count,32949.000000,32949.000000,32949.000000,32949.000000,32341.000000
mean,2276.472488,30.815078,16.937661,23.196728,2.188986
std,4282.038731,16.914458,13.637554,12.079047,1.736766
min,0.000000,7.000000,2.000000,6.000000,1.000000
25%,300.000000,18.000000,8.000000,15.000000,1.000000
50%,700.000000,25.000000,13.000000,20.000000,1.000000
75%,1900.000000,38.000000,21.000000,30.000000,3.000000
max,40425.000000,105.000000,105.000000,118.000000,20.000000


In [35]:
orders_clean['order_status']
customers_clean['customer_city']
customers_clean['customer_state']
products_clean['product_category_name']

0                               perfumaria
1                                    artes
2                            esporte_lazer
3                                    bebes
4                    utilidades_domesticas
                       ...                
32946                     moveis_decoracao
32947    construcao_ferramentas_iluminacao
32948                      cama_mesa_banho
32949               informatica_acessorios
32950                      cama_mesa_banho
Name: product_category_name, Length: 32951, dtype: str

In [36]:
orders_clean['order_status'].nunique(dropna=False)

8

In [37]:
orders_clean['order_status'].unique()

<StringArray>
[  'delivered',    'invoiced',     'shipped',  'processing', 'unavailable',
    'canceled',     'created',    'approved']
Length: 8, dtype: str

In [38]:
customers_clean['customer_city'].nunique(dropna=False)

4119

In [39]:
customers_clean['customer_city'].unique()

<StringArray>
[               'franca', 'sao bernardo do campo',             'sao paulo',
       'mogi das cruzes',              'campinas',        'jaragua do sul',
               'timoteo',              'curitiba',        'belo horizonte',
         'montes claros',
 ...
              'balbinos',          'serra bonita',          'venda branca',
           'sanga puita',               'queiroz',                'siriji',
   'natividade da serra',          'monte bonito',            'sao rafael',
     'eugenio de castro']
Length: 4119, dtype: str

In [40]:
customers_clean['customer_state'].nunique(dropna=False)

27

In [41]:
customers_clean['customer_state'].unique()

<StringArray>
['SP', 'SC', 'MG', 'PR', 'RJ', 'RS', 'PA', 'GO', 'ES', 'BA', 'MA', 'MS', 'CE',
 'DF', 'RN', 'PE', 'MT', 'AM', 'AP', 'AL', 'RO', 'PB', 'TO', 'PI', 'AC', 'SE',
 'RR']
Length: 27, dtype: str

In [42]:
products_clean['product_category_name'].nunique(dropna=False)

74

In [43]:
products_clean['product_category_name'].unique()

<StringArray>
[                                    'perfumaria',
                                          'artes',
                                  'esporte_lazer',
                                          'bebes',
                          'utilidades_domesticas',
                          'instrumentos_musicais',
                                     'cool_stuff',
                               'moveis_decoracao',
                               'eletrodomesticos',
                                     'brinquedos',
                                'cama_mesa_banho',
               'construcao_ferramentas_seguranca',
                         'informatica_acessorios',
                                   'beleza_saude',
                               'malas_acessorios',
                             'ferramentas_jardim',
                              'moveis_escritorio',
                                     'automotivo',
                                    'eletronicos',
                 

In [44]:
orders_clean['order_status'] = (
    orders_clean['order_status']
    .str.strip()
    .str.lower()
)
orders_clean['order_status']

0        delivered
1        delivered
2        delivered
3        delivered
4        delivered
           ...    
99436    delivered
99437    delivered
99438    delivered
99439    delivered
99440    delivered
Name: order_status, Length: 99441, dtype: str

In [45]:
customers_clean['customer_city'] = (
    customers_clean['customer_city']
    .str.strip()
    .str.lower()
)
customers_clean['customer_city']

0                       franca
1        sao bernardo do campo
2                    sao paulo
3              mogi das cruzes
4                     campinas
                 ...          
99436                sao paulo
99437          taboao da serra
99438                fortaleza
99439                   canoas
99440                    cotia
Name: customer_city, Length: 99441, dtype: str

In [46]:
customers_clean['customer_state'] = (
    customers_clean['customer_state']
    .str.strip()
    .str.upper()
)
customers_clean['customer_state']

0        SP
1        SP
2        SP
3        SP
4        SP
         ..
99436    SP
99437    SP
99438    CE
99439    RS
99440    SP
Name: customer_state, Length: 99441, dtype: str

In [47]:
products_clean['product_category_name'] = (
    products_clean['product_category_name']
    .str.strip()
    .str.lower()
)
products_clean['product_category_name']

0                               perfumaria
1                                    artes
2                            esporte_lazer
3                                    bebes
4                    utilidades_domesticas
                       ...                
32946                     moveis_decoracao
32947    construcao_ferramentas_iluminacao
32948                      cama_mesa_banho
32949               informatica_acessorios
32950                      cama_mesa_banho
Name: product_category_name, Length: 32951, dtype: str

In [48]:
orders_clean['order_status'].nunique(dropna=False)


8

In [49]:
customers_clean['customer_city'].nunique(dropna=False)


4119

In [50]:
customers_clean['customer_state'].nunique(dropna=False)

27

In [51]:
products_clean['product_category_name'].nunique(dropna=False)

74

In [52]:
orders_clean.isna().sum()

order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

In [53]:
orders_clean.isna().mean() * 100

order_id                         0.000000
customer_id                      0.000000
order_status                     0.000000
order_purchase_timestamp         0.000000
order_approved_at                0.160899
order_delivered_carrier_date     1.793023
order_delivered_customer_date    2.981668
order_estimated_delivery_date    0.000000
dtype: float64

In [54]:
customers_clean.isna().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [55]:
customers_clean.isna().mean() * 100

customer_id                 0.0
customer_unique_id          0.0
customer_zip_code_prefix    0.0
customer_city               0.0
customer_state              0.0
dtype: float64

In [56]:
products_clean.isna().sum()

product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

In [57]:
products_clean.isna().mean() * 100

product_id                    0.000000
product_category_name         1.851234
product_name_lenght           1.851234
product_description_lenght    1.851234
product_photos_qty            1.851234
product_weight_g              0.006070
product_length_cm             0.006070
product_height_cm             0.006070
product_width_cm              0.006070
dtype: float64

In [58]:
order_items_clean.isna().sum()


order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

In [59]:
order_items_clean.isna().mean() * 100

order_id               0.0
order_item_id          0.0
product_id             0.0
seller_id              0.0
shipping_limit_date    0.0
price                  0.0
freight_value          0.0
dtype: float64

In [60]:
orders_clean['order_id'].duplicated().sum()

np.int64(0)

In [61]:
order_items_clean.duplicated(
    subset = ['order_id', 'order_item_id']
).sum()

np.int64(0)

In [62]:
customers_clean['customer_id'].duplicated().sum()

np.int64(0)

In [63]:
products_clean['product_id'].duplicated().sum()

np.int64(0)

In [64]:
customers_clean.duplicated(
    subset=['customer_unique_id']
).sum()

np.int64(3345)

In [65]:
orders_customers = orders_clean.merge(
    customers_clean,
    on="customer_id",
    how="left",
    validate='many_to_one',
    indicator='customer_match')

In [66]:
print("shape:", orders_customers.shape)
'\n=====  ====='
print(orders_customers['customer_match'].value_counts())
'\n=====  ====='
print(orders_customers['order_id'].nunique())
'\n=====  ====='
print(orders_customers['customer_unique_id'].nunique())

shape: (99441, 13)
customer_match
both          99441
left_only         0
right_only        0
Name: count, dtype: int64
99441
96096


In [67]:
orders_items = orders_clean.merge(
    order_items_clean,
    on="order_id",
    how="left",
    validate='one_to_many',
    indicator='items_match'
)

In [68]:
print("shape:", orders_items.shape)
'\n=====  ====='
print(orders_items['items_match'].value_counts())


shape: (113425, 15)
items_match
both          112650
left_only        775
right_only         0
Name: count, dtype: int64


In [69]:
orders_without_items = orders_items.loc[
    orders_items['items_match'] == 'left_only'
]
orders_without_items.shape

(775, 15)

In [70]:
orders_without_items['order_status'].value_counts()

order_status
unavailable    603
canceled       164
created          5
invoiced         2
shipped          1
Name: count, dtype: int64

In [71]:
items_products = order_items_clean.merge(
    products_clean,
    on="product_id",
    how="left",
    validate='many_to_one',
    indicator= 'product_match'
)

In [72]:
print("shape:", items_products.shape)
'\n=====  ====='
print(items_products['product_match'].value_counts())
'\n=====  ====='
print(items_products['product_id'].nunique())

shape: (112650, 16)
product_match
both          112650
left_only          0
right_only         0
Name: count, dtype: int64
32951


In [73]:
items_products['price'].sum()

np.float64(13591643.7)

In [74]:
items_products['freight_value'].sum()

np.float64(2251909.54)

In [75]:
first_full_data_merge = order_items_clean.merge(
    orders_clean,
    on="order_id",
    how="left",
    validate="many_to_one",
)

In [76]:
second_full_data_merge = first_full_data_merge.merge(
    customers_clean,
    on="customer_id",
    how="left",
    validate="many_to_one",
)

In [77]:
full_data = second_full_data_merge.merge(
    products_clean,
    on="product_id",
    how="left",
    validate="many_to_one",
)

In [78]:

unique_order_id = full_data['order_id'].nunique()

unique_customer_id = full_data['customer_unique_id'].nunique()

totalPrice = full_data['price'].sum()

totalFreight = full_data['freight_value'].sum()

In [79]:
print(len(full_data))

print(unique_order_id)

print(unique_customer_id)

print(totalPrice)

print(totalFreight)

112650
98666
95420
13591643.7
2251909.54


In [80]:
quality_report = pd.DataFrame([
    {
        'table_name': 'orders',
        'row_count': len(orders),
        'column_count': orders.shape[1],
        'duplicate_key_count': orders['order_id'].duplicated().sum(),
        'missing_cells': orders.isna().sum().sum()
    },
    {
        'table_name': 'customers',
        'row_count': len(customers),
        'column_count': customers.shape[1],
        'duplicate_key_count': customers['customer_id'].duplicated().sum(),
        'missing_cells': customers.isna().sum().sum()
    },
    {
        'table_name': 'order_items',
        'row_count': len(order_items),
        'column_count': order_items.shape[1],
        'duplicate_key_count': order_items.duplicated(subset=['order_id', 'order_item_id']).sum(),
        'missing_cells': order_items.isna().sum().sum()
    },
    {
        'table_name': 'products',
        'row_count': len(products),
        'column_count': products.shape[1],
        'duplicate_key_count': products['product_id'].duplicated().sum(),
        'missing_cells': products.isna().sum().sum()
    }
    
])
quality_report

,table_name,row_count,column_count,duplicate_key_count,missing_cells
0,orders,99441,8,0,4908
1,customers,99441,5,0,0
2,order_items,112650,7,0,0
3,products,32951,9,0,2448


-unique are keys: orders.order_idm customers.customer_id, products.product_id and in order_items there is complex key (order_id, order_item_id)


-

### Выводы по качеству данных

1. Ключи `orders.order_id`, `customers.customer_id` и `products.product_id` уникальны. В `order_items` уникальна комбинация `(order_id, order_item_id)`. Полных дубликатов ключей не обнаружено. При этом `customer_unique_id` повторяется 3 345 раз, но это не ошибка: один реальный клиент может оформить несколько заказов.

2. В таблицах `customers` и `order_items` пропусков нет. В `orders` обнаружено 4 908 пропущенных ячеек: 160 в `order_approved_at`, 1 783 в `order_delivered_carrier_date` и 2 965 в `order_delivered_customer_date`.

3. В `products` обнаружено 2 448 пропущенных ячеек. По 610 пропусков содержат `product_category_name`, `product_name_lenght`, `product_description_lenght` и `product_photos_qty`. В столбцах веса, длины, высоты и ширины товара обнаружено по два пропуска.

4. Большинство отсутствующих дат доставки являются ожидаемыми для заказов со статусами `canceled`, `unavailable`, `shipped`, `invoiced`, `processing` и `created`, поскольку такие заказы не были полностью доставлены. Отсутствие даты доставки у заказа со статусом `delivered` следует считать потенциальной проблемой качества данных.

5. Обнаружено 775 заказов без позиций: 603 со статусом `unavailable`, 164 — `canceled`, 5 — `created`, 2 — `invoiced` и 1 — `shipped`. Большинство таких случаев связано с отменёнными или недоступными заказами, которые не дошли до нормального завершения. Оставшиеся случаи могут указывать на неполноту исходных данных.

6. Все заказы успешно сопоставились с таблицей клиентов, а все позиции заказов — с таблицей товаров. Проверки `validate` подтвердили ожидаемые связи `one-to-many` и `many-to-one`.

7. Результаты Pandas полностью совпали с PostgreSQL: 112 650 строк, 98 666 уникальных заказов, 95 420 реальных клиентов, общая стоимость товаров 13 591 643,70 и стоимость доставки 2 251 909,54.


In [88]:
def table_quality(df, table_name, key_columns):
    return {
        'table_name' : table_name,
        'row_count' : len(df),
        'column_count' : df.shape[1],
        'duplicate_key_count' : df.duplicated(subset=key_columns).sum(),
        'missing_cells' : df.isna().sum().sum()
    }

In [89]:
table_quality(orders, 'orders', ['order_id'])

{'table_name': 'orders',
 'row_count': 99441,
 'column_count': 8,
 'duplicate_key_count': np.int64(0),
 'missing_cells': np.int64(4908)}

In [90]:
quality_report_function = pd.DataFrame([
    table_quality(orders, 'orders', ['order_id']),
    table_quality(customers, 'customers', ['customer_id']),
    table_quality(order_items, 'order_items', ['order_id', 'order_item_id']),
    table_quality(products, 'products', ['product_id'])
])

quality_report_function

,table_name,row_count,column_count,duplicate_key_count,missing_cells
0,orders,99441,8,0,4908
1,customers,99441,5,0,0
2,order_items,112650,7,0,0
3,products,32951,9,0,2448
